yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -207,6 +211,17 @@ private[yarn] class YarnAllocator(
   }
 
   /**
+   * Gets the executor loss reason for a disconnected executor.
+   * Note that this method is expected to be called exactly once per executor ID.
+   */
+  def getExecutorLossReason(executorId: String): ExecutorLossReason = synchronized {
+    allocateResources()
+    // Expect to be asked for a loss reason once and exactly once.
+    assert(completedExecutorExitReasons.contains(executorId))
```

Comment:
In such a case that AMRM client can't be guaranteed to find the latest container exit statuses, then we would need to do polling and perhaps "give up" after a few tries? I don't see how we could ever guarantee that getExecutorLossReason could ever be correct if we can't ever be guaranteed to find the correct state from the NM.